C/services/north/data_send.cpp

```diff
@@ -248,3 +268,70 @@ void DataSender::releasePause()
 	}
 	m_pauseCV.notify_all();
 }
+
+/**
+ * Update the sent statistics
+ *
+ * @param increment     Increment of the number of readings sent
+ */
+void DataSender::updateStatistics(uint32_t increment)
+{
+	lock_guard<mutex> guard(m_statsMtx);
+
+	// Add statistics counter to the map
+	m_statsPendingEntries[m_loader->getName()] += increment;
+	m_statsPendingEntries["Readings Sent"] += increment;
+}
+
+/**
+ * Flush statistics top storage service
+ */
+void DataSender::flushStatistics()
+{
+	if (m_shutdown)
+	{
+		return;
+	}
+
+	// Wait for FLUSH_STATS_INTERVAL seconds or receive notification
+	// when shutdown is called
+	unique_lock<mutex> flush(m_flushStatsMtx);
+	m_statsCv.wait_for(flush, std::chrono::seconds(FLUSH_STATS_INTERVAL));
+	flush.unlock();
+
+	std::map<std::string, int> statsData;
+
+	// Acquire m_statsMtx lock for m_statsMtx
+	unique_lock<mutex> lck(m_statsMtx);
+
+	// copy statistics map
+	statsData = m_statsPendingEntries;
+
+	// Reset statistics
+	m_statsPendingEntries.clear();
+
+	// Release lock
+	lck.unlock();
+
+	if (statsData.empty())
+	{
+		return;
+	}
+
+	// Send statistics to storage service
+	map<string, int>::iterator it;
+	for (it = statsData.begin(); it != statsData.end(); it++)
+	{
+		string description;
+		if (it->first == m_loader->getName())
+		{
+			description = it->first + " Readings Sent";
+		}
+		else
+		{
+			description = it->first + " Noth";
+		}
+		m_loader->updateStatistic(it->first, description, it->second);
```

Comment:
No, this is wrong, you are still making multiple calls to the storage layer. The idea is to make one call to the storage layer with all the statistics to update.